### ISSUE TITLE
'platform_type' option in avh.query fails

### ISSUE THREAD

**Issue by SBS-EREHM** [argovis/argovis_helpers#16]
The query key 'platform_type'  (with value from [NVS Table R23](https://vocab.nerc.ac.uk/collection/R23/current/)  'Argo platform type' ) crashes with a JSONDecode Error.    (argovisHelpers 0.0.10)

Example:  When I run the query below, it crashes when provided with a platform type (PROVOR_III) that I know has profiles within the selected polygon.  (Note that this query executes without error if I remove the 'platform_type' option from `params`.

```
dataSetName='NE Labrador Sea'
startDate = '2010-01-01T00:00:00Z'  
polygon = '[[-55.02669685509557,59.72456188338252],[-55.02669685509557,56.91050178921344],[-49.75400577229302,57.05417572583986],[-50.193396695859896,59.90133567536662],[-55.02669685509557,59.72456188338252]]'  # LabradorSeaNE
startDate = '2018-10-13T00:00:00Z'
nowiso = datetime.utcnow().replace(microsecond=0).isoformat()
params = {
        'startDate': startDate,
        'endDate':   nowiso+'Z',
        'source': 'argo_bgc',
        'polygon': eval(polygon),
        'data': 'all',
        'platform_type' : 'PROVOR_III'
    }
print(dataSetName)
print(params)
# Make the query. Returns a list of JSON instances, one per profile
dd = avh.query('argo', options=params, apikey=API_KEY, apiroot=API_PREFIX)
```
It crashes with a **JSONDecode** error followed by a 504 Gateway Time-out

```
JSONDecodeError                           Traceback (most recent call last)
File [/opt/anaconda3/envs/sbs39/lib/python3.10/site-packages/requests/models.py:910](https://file+.vscode-resource.vscode-cdn.net/opt/anaconda3/envs/sbs39/lib/python3.10/site-packages/requests/models.py:910), in Response.json(self, **kwargs)
    909 try:
--> 910     return complexjson.loads(self.text, **kwargs)
    911 except JSONDecodeError as e:
    912     # Catch JSON-related errors and raise as requests.JSONDecodeError
    913     # This aliases json.JSONDecodeError and simplejson.JSONDecodeError

File [/opt/anaconda3/envs/sbs39/lib/python3.10/json/__init__.py:346](https://file+.vscode-resource.vscode-cdn.net/opt/anaconda3/envs/sbs39/lib/python3.10/json/__init__.py:346), in loads(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    343 if (cls is None and object_hook is None and
    344         parse_int is None and parse_float is None and
    345         parse_constant is None and object_pairs_hook is None and not kw):
--> 346     return _default_decoder.decode(s)
    347 if cls is None:

File [/opt/anaconda3/envs/sbs39/lib/python3.10/json/decoder.py:337](https://file+.vscode-resource.vscode-cdn.net/opt/anaconda3/envs/sbs39/lib/python3.10/json/decoder.py:337), in JSONDecoder.decode(self, s, _w)
    333 """Return the Python representation of ``s`` (a ``str`` instance
    334 containing a JSON document).
    335 
    336 """
--> 337 obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338 end = _w(s, end).end()

File [/opt/anaconda3/envs/sbs39/lib/python3.10/json/decoder.py:355](https://file+.vscode-resource.vscode-cdn.net/opt/anaconda3/envs/sbs39/lib/python3.10/json/decoder.py:355), in JSONDecoder.raw_decode(self, s, idx)
...

JSONDecodeError: [Errno Expecting value] 
504 Gateway Time-out

The server didn't respond in time.

: 0
```

**Comment by SBS-EREHM** [comment]
Note that if you change startDate = '2014-01-01T00:00:00Z' , the query succeeds with platform_type set to 'PROVOR_III'.

**Comment by katieannemills** [comment]
Thanks for reaching out here - the following, which is exactly the same as your example above, runs for me:

```
import argovisHelpers.helpers as avh
import datetime


dataSetName='NE Labrador Sea'
polygon = '[[-55.02669685509557,59.72456188338252],[-55.02669685509557,56.91050178921344],[-49.75400577229302,57.05417572583986],[-50.193396695859896,59.90133567536662],[-55.02669685509557,59.724561883>
startDate = '2018-10-13T00:00:00Z'
nowiso = datetime.datetime.utcnow().replace(microsecond=0).isoformat()
params = {
        'startDate': startDate,
        'endDate':   nowiso+'Z',
        'source': 'argo_bgc',
        'polygon': eval(polygon),
        'data': 'all',
        'platform_type' : 'PROVOR_III'
    }
print(dataSetName)
print(params)
# Make the query. Returns a list of JSON instances, one per profile
dd = avh.query('argo', options=params, apikey='guest', apiroot='https://argovis-api.colorado.edu/')
print(dd)
```

Are you still seeing the same error? What version of argovisHelpers do you have installed, we're on 0.0.10 atm.

**Comment by SBS-EREHM** [comment]
OK.  Works now for me too.  (FYI, I have been using argovisHelpers 0.0.10)

But now change startDate as follows:
````startDate = '2012-10-13T00:00:00Z' ````

I get the same '504 Gateway Time-out' again after approx 5.5 minutes of waiting for the query to come back....
Are my queries simply taking too long?  Is there a way I can lengthen this time-out period?

Of course, I can try breaking it up the long period query into smaller queries and concatenate the results but ....

Thanks,

/eric

**Comment by SBS-EREHM** [comment]
Some more tests:

Try the same query as above, replacing startDate and endDate option parameters with:
````
startDate = '2012-10-13T00:00:00Z'  
endDate =   '2016-10-13T00:00:00Z'  
````
This works for me, returning 471 profiles in 50 s.  

But then simply increment the endDate by one year:
````
startDate = '2012-10-13T00:00:00Z'  
endDate =   '2017-10-13T00:00:00Z'  
````
and  I get  again the same  '504 Gateway Time-out'  again after about 5 min.

So I try the query:
````
startDate = '2016-10-13T00:00:00Z'  
endDate =   '2017-10-13T00:00:00Z'  
````
and it completes successfully in 14 s, with 46 profiles returned.

This is reproducible at will for me.  Let me know if you can reproduce this.

**Comment by katieannemills** [comment]
Thanks for this @SBS-EREHM - I've been playing with this all day and get qualitatively similar results to you; server side compute consumption appears nonlinear in time interval requested, likely due to deserialization of a lot of BGC measurements all at once. I just released argovisHelpers 0.0.11, which will slice long timeranges automatically for you. I can make your original request successfully in under a minute using that update.

_Edit: your original request, but starting in 2012._

**Comment by katieannemills** [comment]
To further support BGC consumers, https://github.com/argovis/argovis_api/releases/tag/2.10.2 significantly improves API performance for BGC requests that select a small subset of available measurement variables using the `data` option - not strictly applicable to the above `data=all` example, but nonetheless useful when targeting specific BGC measurements is interesting.

Closing out here as I think we've addressed this at the API and helper package levels, but please add further comments or more issues as needed.
